content-tracker/src/main.rs

```diff
@@ -0,0 +1,261 @@
+pub mod args;
+pub mod discovery;
+pub mod io;
+pub mod iroh_bytes_util;
+pub mod options;
+pub mod protocol;
+pub mod tracker;
+
+use std::{
+    collections::BTreeSet,
+    sync::atomic::{AtomicBool, Ordering},
+    time::{Duration, Instant},
+};
+
+use clap::Parser;
+use io::CONFIG_DEFAULTS_FILE;
+use iroh::util::fs::load_secret_key;
+use iroh_net::{
+    magic_endpoint::{get_alpn, get_peer_id},
+    AddrInfo, MagicEndpoint, NodeAddr,
+};
+use tokio_util::task::LocalPoolHandle;
+
+use crate::{
+    args::{AnnounceArgs, Args, Commands, QueryArgs, ServerArgs},
+    io::{load_from_file, setup_logging, tracker_home, tracker_path, CONFIG_FILE, SERVER_KEY_FILE},
+    options::Options,
+    protocol::{
+        Announce, AnnounceKind, Query, QueryFlags, Request, Response, REQUEST_SIZE_LIMIT,
+        TRACKER_ALPN,
+    },
+    tracker::Tracker,
+};
+
+pub type NodeId = iroh_net::key::PublicKey;
+
+static VERBOSE: AtomicBool = AtomicBool::new(false);
+
+fn set_verbose(verbose: bool) {
+    VERBOSE.store(verbose, Ordering::Relaxed);
+}
+
+pub fn verbose() -> bool {
+    VERBOSE.load(Ordering::Relaxed)
+}
+
+#[macro_export]
+macro_rules! log {
+    ($($arg:tt)*) => {
+        if $crate::verbose() {
+            println!($($arg)*);
+        }
+    };
+}
+
+/// Wait until the endpoint has figured out it's own DERP region.
+async fn await_derp_region(endpoint: &MagicEndpoint) -> anyhow::Result<()> {
+    let t0 = Instant::now();
+    loop {
+        let addr = endpoint.my_addr().await?;
+        if addr.derp_region().is_some() {
+            break;
+        }
+        if t0.elapsed() > Duration::from_secs(10) {
+            anyhow::bail!("timeout waiting for DERP region");
+        }
+        tokio::time::sleep(Duration::from_millis(50)).await;
+    }
+    Ok(())
+}
+
+async fn create_endpoint(
+    key: iroh_net::key::SecretKey,
+    port: u16,
+) -> anyhow::Result<MagicEndpoint> {
+    // let pkarr_relay_discovery = discovery::PkarrRelayDiscovery::new(key.clone(), PKARR_RELAY_URL.parse().unwrap());
+    let region_discover = discovery::HardcodedRegionDiscovery::new(2);
+    iroh_net::MagicEndpoint::builder()
+        .secret_key(key)
+        .discovery(Box::new(region_discover))
+        .alpns(vec![TRACKER_ALPN.to_vec()])
+        .bind(port)
+        .await
+}
+
+/// Accept an incoming connection and extract the client-provided [`NodeId`] and ALPN protocol.
+pub async fn accept_conn(
+    mut conn: quinn::Connecting,
+) -> anyhow::Result<(NodeId, String, quinn::Connection)> {
+    let alpn = get_alpn(&mut conn).await?;
+    tracing::info!("awaiting conn");
+    let conn = conn.await?;
+    tracing::info!("got conn");
```

Comment:
include the peer_id in the log?

Reply:
this is done further down. Removing this.